src/scheduler/gc_work.rs

```diff
@@ -223,6 +223,11 @@ impl<VM: VMBinding> GCWork<VM> for EndOfGC {
     fn do_work(&mut self, worker: &mut GCWorker<VM>, mmtk: &'static MMTK<VM>) {
         info!("End of GC");
 
+        // We assume this is the only running work packet that accesses plan at the point of execution
+        #[allow(clippy::cast_ref_to_mut)]
+        let plan_mut: &mut dyn Plan<VM = VM> = unsafe { &mut *(&*mmtk.plan as *const _ as *mut _) };
```

Comment:
I see this everywhere -- maybe we should just change the plan to `mut Plan` or something in the `MMTK` struct. Or am I missing something which means we can't do that?

Reply:
This is the same pattern as our prepare/release code. This is unfortunate. 
https://github.com/mmtk/mmtk-core/blob/2300ce1d4a7c99053c5bb7055b377a71c9241e85/src/scheduler/gc_work.rs#L44-L46
https://github.com/mmtk/mmtk-core/blob/2300ce1d4a7c99053c5bb7055b377a71c9241e85/src/scheduler/gc_work.rs#L113-L115

We can fix it, but not as simple as making it `mut mmtk`. Basically `mmtk: &'static MMTK<VM>` is a non mutable reference, but we want to get a mutable reference from it -- we do so under the assumption that we are the only thread that is accessing `mmtk`. As long as the assumption holds, the code is correct.

Basically what we want is something like [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html) but without overhead on borrowing. I don't think there is a safe way to do it.

Reply:
Why is it not something like an `Arc`? Multiple work packets should be executing at the same time, all of them accessing the `mmtk` variable. Overhead of reference counting? I guess if 99% of the times it's only read like most work packets, and this is the hot loop, ugly hacks like this may be required.

Reply:
Technically a `RWLock` could help, but I feel like adding a lock in the hottest part of our code may not be ideal.

Reply:
`Arc` can help manage the lifetime of plan/mmtk. But it comes at the cost of a fat pointer (unknown cost) and reference counting (which should be minimal cost). But it does not solve the issue of mutability.

There are many things that we need to tidy up.

Reply:
Ah right sorry -- you need a `RefCell` or `Mutex`/`RWLock` to have mutability. Sorry was thinking of `Arc<Mutex<T>>` which is a probably a bad idead.